impeller/entity/contents/filters/gaussian_blur_filter_contents.cc

```diff
@@ -200,6 +207,94 @@ Rect MakeReferenceUVs(const Rect& reference, const Rect& rect) {
 int ScaleBlurRadius(Scalar radius, Scalar scalar) {
   return static_cast<int>(std::round(radius * scalar));
 }
+
+class GeometryExtractor : public TypedContentsVisitor<ColorSourceContents> {
+ public:
+  void TypedVisit(ColorSourceContents* contents) override {
+    geometry_ = contents->GetGeometry();
+  }
+  std::shared_ptr<Geometry> geometry_;
+};
+
+std::shared_ptr<Geometry> GetGeometry(
+    const std::shared_ptr<FilterInput>& input) {
```

Comment:
Okay, I've replaced it with a `FML_DCHECK(false) << "unimplemented"`

Reply:
The problem is that `Texture` inputs are a possible input that doesn't work. We're going to forget this is here and crash users. This is why FilterInput exists: It's an abstraction so we don't have to test every kind of input with every kind of filter.

Reply:
It's not going to crash users.  It's a DCHECK and not having a geometry functions as having no clip.  Also, mask blur's don't work with textures yet, so you can't even get down this path.  There is a follow up issue to address that: https://github.com/flutter/flutter/issues/144266

Reply:
It'll crash Flutter devs if we start using `FilterInput::Make(texture, transform)` in Aiks for arbitrary filters (something we used to do). Often times when we modify a path involving arbitrary/unknown filters, we'll add a test that uses a blend filter (which before this change was sufficient because `FilterInput` is opaque), so there's a good chance we'd miss this.

One way to address this concern would be to implement the missing paths and test the code, but the same would go for every other usage of `FilterInput::GetInput()` within a filter which is why I'm concerned about this: `FilterInput` is an adapter. Allowing filters to disambiguate the underlying input type isn't a good fit.

I added the virtual `FilterInput::GetInput()`, but that was a mistake and it should be removed IMO.

Reply:
> It'll crash Flutter devs if we start using FilterInput::Make(texture, transform) in Aiks for arbitrary filters (something we used to do).

If we do that the test runner will start crashing which will signal this needs to be addressed.  Someone can easily transition to a print statement if they want to send people down this route and have it operate as a noop.

> One way to address this concern would be to implement the missing paths and test the code

That sounds good.  That doesn't have happen in this PR.  We don't block PRs because they haven't implemented everything. If that was the case we'd never have submitted anything =)

Until we've implemented all functionality there will always be a frontier.  In this case I've explicitly listed it and added a debug assert.  I think that's preferable to hiding it and silently noop'ing.

> I added the virtual FilterInput::GetInput(), but that was a mistake and it should be removed IMO.

That isn't the state of the code now though, that refactor can consider this usage if it ever happens.  We can't design code around all the different options people have in their head.

Reply:
Concretely, would it address your concerns if we made this a DLOG instead of a DCHECK?  I think a DCHECK is preferable but sounds like you'd be more comfortable with a DLOG.

Reply:
> That sounds good. That doesn't have happen in this PR. We don't block PRs because they haven't implemented everything. If that was the case we'd never have submitted anything =)

That's not why I didn't approve. Right after this I wrote:
> but the same would go for every other usage of FilterInput::GetInput() within a filter. That's why I'm concerned about this: FilterInput is an adapter. Allowing filters to disambiguate the underlying input type isn't a good fit.

My concern is about the design.

Reply:
> Why should we branch on the variant in a filter implementation instead of FilterInput::GetGeometry?

Did you get a chance to look through the thread I had with Jonah?  He had similar questions about `Contents::GetGeometry`.  Here is my response: https://github.com/flutter/engine/pull/51018#discussion_r1507955149

The problem is that `FilterInput::GetGeometry` does not have a definitive result because it is predicated on its Contents.  And a `Contents` does not have a definitive answer for `GetGeometry` because it's nonsensical for a subset of the implementers of `Contents`.

This solution meets the code where it is today and is the proper way to implement this functionality given the current state of the API.  If we wanted to introduce `FilterInput::GetGeometry` that would require a larger refactor that would require it to be apropos for all FilterInputs.

Reply:
I don't really love the visitor pattern either TBH. But the visitor pattern is different than reaching into a std::variant like this. I agree with Brandon that this should be a method on FilterInput.

Reply:
Notes from talking in person:  The question about passing in the Geometry to the blur was raised again, Brandon missed that conversation the first time and he was in agreement with Jonah that that makes sense.

I objected because implementing things that way is ignoring the Geometry attached to the FilterInput which leads to a whole section of the input domain for the blur to be essentially undefined behavior.  Jonah's point there is that the Canvas is enforcing a limit to the input of the blur, and we only care about that.  I objected again saying that relying on the knowledge of how something is used when implementing it and testing it makes a less robust and tested system; each component should be able to be tested and verified in isolation, ignoring how it is actually used.

I conceded to pass in the Geometry in under the grounds that Brandon and Jonah both found that to be agreeable and in alignment with their vision of how impeller should be developed.  I found that vision just suboptimal, not wrong.

Reply:
I started making the change to pass in the Geometry and unfortunately it doesn't make sense in all cases.  The first oddity I ran into is this:
```c++
std::shared_ptr<FilterContents> BlurImageFilter::WrapInput(
    const FilterInput::Ref& input) const {
  return FilterContents::MakeGaussianBlur(input, sigma_x_, sigma_y_,
                                          blur_style_, tile_mode_);
}
```

If we continue down the route of passing in geometry we'd have to do 
```c++
std::shared_ptr<FilterContents> BlurImageFilter::WrapInput(
    const FilterInput::Ref& input, const std::shared_ptr<Geometry>& geometry) const {
  return FilterContents::MakeGaussianBlur(input, sigma_x_, sigma_y_, geometry,
                                          blur_style_, tile_mode_);
}
```

What do you 2 think?

Reply:
I think that SGTM

Reply:
Notes from chat: In this case we believe `BlurImageFilter::WrapInput` can safely pass nullptr to geometry since the geometry is only used for mask blur filters and mask blur filters should never be going down this path.

Reply:
We should add an assert too to make sure this path isn't used for mask blurs.

Reply:
Also I had forgotten that we only need to worry about the blur style for mask blurs. Mask blurs are set via `Paint::mask_blur_descriptor`, which has its own type distinct from `ImageFilter`. So just the two `FilterContents::MakeGaussianBlur` invocations within `Paint::MaskBlurDescriptor::CreateMaskBlur` would need to be changed.